src/shogun-core-main/src/main/java/de/terrestris/shoguncore/util/interceptor/secure/WfsRequestInterceptor.java

```diff
@@ -0,0 +1,247 @@
+
+    /**
+     *
+     */
+    private URI appUri = null;
+
+    /**
+     *
+     */
+    private static final Logger LOG = getLogger(WfsRequestInterceptor.class);
+
+    /**
+     *
+     */
+    @Autowired
+    @Qualifier("userService")
+    protected UserService<User, UserDao<User>> userService;
+
+    /**
+     *
+     */
+    @Autowired
+    @Qualifier("layerService")
+    protected LayerService<Layer, LayerDao<Layer>> layerService;
+
+    /**
+     *
+     */
+    @Value("${geoserver.username:}")
+    private String gsUser;
+
+    /**
+     *
+     */
+    @Value("${geoserver.password:}")
+    private String gsPass;
+
+   /**
+     *
+     */
+    @Override
+    public MutableHttpServletRequest interceptGetCapabilities(MutableHttpServletRequest request) {
+        LOG.debug("Intercepting WFS GetCapabilities");
+        return new GeoserverAuthHeaderRequest(request, gsUser, gsPass);
+    }
+
+    /**
+     *
+     */
+    @Override
+    public MutableHttpServletRequest interceptDescribeFeatureType(MutableHttpServletRequest request) {
+        LOG.debug("Intercepting WFS DescribeFeatureType");
+        return new GeoserverAuthHeaderRequest(request, gsUser, gsPass);
+    }
+
+    /**
+     *
+     */
+    @Override
+    public MutableHttpServletRequest interceptGetFeature(MutableHttpServletRequest request) {
+        LOG.debug("Intercepting WFS GetFeature");
+        return isAllowed(request, "READ") ? new GeoserverAuthHeaderRequest(
+            request, gsUser, gsPass) : forbidRequest(request);
+    }
+
+    /**
+     *
+     */
+    @Override
+    public MutableHttpServletRequest interceptLockFeature(MutableHttpServletRequest request) {
+        LOG.debug("Intercepting WFS LockFeature");
+        return isAllowed(request, "UPDATE") ? new GeoserverAuthHeaderRequest(
+            request, gsUser, gsPass) : forbidRequest(request);
+    }
+
+    /**
+     *
+     */
+    @Override
+    public MutableHttpServletRequest interceptTransaction(MutableHttpServletRequest request) {
+        LOG.debug("Intercepting WFS Transaction");
+        return isAllowed(request, "UPDATE") ? new GeoserverAuthHeaderRequest(
+            request, gsUser, gsPass) : forbidRequest(request);
+    }
+
+    /**
+     * Check for permission on layer by getting it through service.
+     * For update permissions, we check through custom method.
+     *
+     * @param request   The request
+     * @param paramName The optional parameter name for the layer parameter
+     * @param method    The access method like read or write
+     * @return if the layer is allowed to read for current user
+     */
+    private boolean isAllowed(
+        MutableHttpServletRequest request, String paramName, String method) {
+        String typeNameParam = request.getParameterIgnoreCase(paramName);
+        List<Layer> all = layerService.findAll();
+        boolean match = false;
+        for (Layer layer : all) {
```

Comment:
Just wondering, but do we really need to iterate all layers? Can't we just get the requested layer/source?

And I would love to have a solution that integrates a cache for the permission as this will _really_ slow down the request (espacially for WMS GetMap).

Reply:
Sadly we can't. The layerservice and more specific layersourceservice cannot be used to find the specific source, as the `findBySimpleFilter` or `findAllWhereFieldEquals` do have varying fields in their implementing child classes. Those methods will return nothing or exceptions when using them e.g. with "layerNames" or "typeNames" filter.

And yes, performance can definately be enhanced, but i need to get this done before and would like to see this enhanced in another PR